src/dataset.rs

```diff
@@ -1,60 +1,112 @@
+use gdal_sys::{
+    self, CPLErr, GDALAccess, GDALDatasetH, GDALMajorObjectH, OGRLayerH, OGRwkbGeometryType,
+};
 use libc::{c_double, c_int};
-use std::ffi::CString;
-use std::path::Path;
-use std::ptr::null_mut;
-
-#[cfg(feature = "ndarray")]
-use ndarray::Array2;
+use ptr::null_mut;
 
 use crate::errors::*;
 
 pub type GeoTransform = [c_double; 6];
+static START: Once = Once::new();
 
 pub struct Dataset {
     c_dataset: GDALDatasetH,
 }
 
-impl MajorObject for Dataset {
-    unsafe fn gdal_object_ptr(&self) -> GDALMajorObjectH {
-        self.c_dataset
+pub fn _register_drivers() {
+    unsafe {
+        START.call_once(|| {
+            gdal_sys::GDALAllRegister();
+        });
     }
 }
 
-impl Metadata for Dataset {}
+impl Dataset {
+    pub unsafe fn from_c_dataset(c_dataset: GDALDatasetH) -> Dataset {
+        Dataset { c_dataset }
+    }
 
-impl Drop for Dataset {
-    fn drop(&mut self) {
-        unsafe {
-            gdal_sys::GDALClose(self.c_dataset);
-        }
+    pub unsafe fn c_dataset(&self) -> GDALDatasetH {
+        self.c_dataset
     }
-}
 
-impl Dataset {
     pub fn open(path: &Path) -> Result<Dataset> {
+        Self::open_ex(path, None, None, None, None)
+    }
+
+    pub fn open_ex(
+        path: &Path,
+        open_flags: Option<u32>,
+        _allowed_drivers: Option<&str>, // TODO: use parameters
+        _open_options: Option<&str>,
+        _sibling_files: Option<&str>,
+    ) -> Result<Dataset> {
         _register_drivers();
         let filename = path.to_string_lossy();
         let c_filename = CString::new(filename.as_ref())?;
-        let c_dataset = unsafe { gdal_sys::GDALOpen(c_filename.as_ptr(), GDALAccess::GA_ReadOnly) };
+        let c_open_flags = open_flags.unwrap_or(GDALAccess::GA_ReadOnly); // This defaults to GdalAccess::GA_ReadOnly
+
+        let c_dataset = unsafe {
+            gdal_sys::GDALOpenEx(
+                c_filename.as_ptr(),
+                c_open_flags,
+                ptr::null(),
+                ptr::null(),
+                ptr::null(),
+            )
+        };
         if c_dataset.is_null() {
-            Err(_last_null_pointer_err("GDALOpen"))?;
+            Err(_last_null_pointer_err("GDALOpenEx"))?;
         }
         Ok(Dataset { c_dataset })
     }
 
-    pub unsafe fn _with_c_ptr(c_dataset: GDALDatasetH) -> Dataset {
+    pub unsafe fn from_c_ptr(c_dataset: GDALDatasetH) -> Dataset {
         Dataset { c_dataset }
     }
 
-    pub unsafe fn _c_ptr(&self) -> GDALDatasetH {
-        self.c_dataset
+    pub fn projection(&self) -> String {
+        let rv = unsafe { gdal_sys::GDALGetProjectionRef(self.c_dataset) };
+        _string(rv)
+    }
+
+    pub fn set_projection(&self, projection: &str) -> Result<()> {
+        let c_projection = CString::new(projection)?;
+        unsafe { gdal_sys::GDALSetProjection(self.c_dataset, c_projection.as_ptr()) };
+        Ok(())
+    }
+
+    pub fn create_copy(&self, driver: &Driver, filename: &str) -> Result<Dataset> {
```

Comment:
```suggestion
    pub fn create_copy(&self, driver: &Driver, filename: &Path) -> Result<Dataset> {
```

Maybe use  a `Path` here like in `open_ex`

Reply:
I think all "paths" in the lib should use  Path instead of &str. I guess thats a new issue.